cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs

```diff
@@ -14,28 +20,98 @@ module Cardano.Crypto.Libsodium.Memory.Internal (
+  pushAllocLogEvent,
+  withAllocLog,
 ) where
 
-import Control.Exception (bracket)
+import Control.Concurrent.STM (atomically)
+import Control.Concurrent.STM.TChan (newTChanIO, TChan, tryReadTChan, writeTChan)
+import Control.Concurrent.STM.TVar (newTVarIO, TVar, readTVar, modifyTVar)
+import Control.DeepSeq (NFData (..), rwhnf)
+import Control.Exception (bracket, bracket_)
 import Control.Monad (when)
 import Data.Coerce (coerce)
 import Data.Proxy (Proxy (..))
 import Foreign.C.Error (errnoToIOError, getErrno)
 import Foreign.C.Types (CSize (..))
-import Foreign.ForeignPtr (ForeignPtr, newForeignPtr, withForeignPtr, finalizeForeignPtr)
-import Foreign.Ptr (Ptr, nullPtr)
+import Foreign.Ptr (Ptr, nullPtr, WordPtr, ptrToWordPtr)
+import Foreign.ForeignPtr (ForeignPtr, withForeignPtr, finalizeForeignPtr, castForeignPtr)
+import Foreign.Concurrent (newForeignPtr)
 import Foreign.Storable (Storable (alignment, sizeOf, peek))
+import Foreign.Marshal.Utils (fillBytes)
 import GHC.TypeLits (KnownNat, natVal)
 import GHC.IO.Exception (ioException)
 import NoThunks.Class (NoThunks, OnlyCheckWhnfNamed (..))
+import System.IO.Unsafe (unsafePerformIO)
 
 import Cardano.Foreign
 import Cardano.Crypto.Libsodium.C
+import Cardano.Memory.Pool (initPool, grabNextBlock, Pool)
+
+-- | Allocation log event. These are emitted automatically whenever mlocked
+-- memory is allocated through the 'allocMLockedForeignPtr' primitive, or
+-- released through an associated finalizer (either explicitly or due to GC).
+-- Additional events that are not actual allocations/deallocations, but may
+-- provide useful debugging context, can be inserted as 'MarkerEv'.
+data AllocEvent
+  = AllocEv !WordPtr
+  | FreeEv !WordPtr
+  | MarkerEv !String
+  deriving (Eq, Show)
+
+{-#NOINLINE allocLog #-}
+allocLog :: TChan AllocEvent
+allocLog = unsafePerformIO newTChanIO
+
+{-#NOINLINE allocLogEnableCounter #-}
+allocLogEnableCounter :: TVar Int
+allocLogEnableCounter = unsafePerformIO $ newTVarIO 0
+
+enableAllocLog :: IO ()
+enableAllocLog = atomically $ modifyTVar allocLogEnableCounter succ
+
+disableAllocLog :: IO ()
+disableAllocLog = atomically $ modifyTVar allocLogEnableCounter pred
+
+drainAllocLog :: IO [AllocEvent]
+drainAllocLog =
+  reverse <$> go []
+  where
+    go xs = do
+      popAllocLogEvent >>= \case
+        Nothing ->
+          return xs
+        Just x ->
+          go (x:xs)
+
+-- | Run an IO action with allocation logging. The allocation log for the
+-- action will be returned as a list of events, in ascending chronological
+-- order.
+withAllocLog :: IO () -> IO [AllocEvent]
+withAllocLog a =
+  bracket_
+    enableAllocLog
+    disableAllocLog
+    (a >> drainAllocLog)
+
+popAllocLogEvent :: IO (Maybe AllocEvent)
+popAllocLogEvent = atomically $ tryReadTChan allocLog
+
+pushAllocLogEvent :: AllocEvent -> IO ()
+pushAllocLogEvent ev = atomically $ do
+  count <- readTVar allocLogEnableCounter
+  when (count > 0) $
+    writeTChan allocLog ev
```

Comment:
The usual logging approaches would be very awkward here; passing a logger object all the way down to the allocator, on the off chance that we might be running one of the few tests that actually need this, would be a lot of clutter for little gain.

The eventual plan is to use `io-sim` for this - all the KES and DSIGNM code is already written against the `MonadSodium` typeclass and a few other `MonadSomething` typeclasses, rather than against `IO` directly, so in theory this should be perfectly feasible, and we can inject the logging into the io-sim machinery without touching the tested code at all. 

Making everything fully io-sim-compatible is a bit of an elaborate task though, and I don't want to do this in this PR just yet, but we can go halfway, making a lightweight wrapper around `IO`, which forwards everything to `IO` itself, except that mlocked allocation calls are intercepted and logged.

Reply:
See other remark; I have removed all the logging stuff from cardano-crypto-class, and rewrote it such that we can inject logging via custom monad stacks (and, if so desired, the `MonadSodium` typeclass). All that now happens inside the test suite.